etk-asm/src/ingest.rs

```diff
@@ -403,70 +283,87 @@ where
+        let nodes = parse_asm(src).context(error::Parse {
+            path: program.sources.last().unwrap().clone(),
+        })?;
+        let mut raws = Vec::new();
+        for node in nodes {
             match node {
                 Node::Op(op) => {
-                    self.sources.write(RawOp::Op(op))?;
+                    raws.push(RawOp::Op(op));
                 }
-                Node::Raw(raw) => {
-                    self.sources.write(RawOp::Raw(raw))?;
+                Node::Import(imp_path) => {
+                    let new_raws = self.resolve_and_ingest(program, imp_path)?;
+                    raws.extend(new_raws);
                 }
-                Node::Import(path) => {
-                    let partial = self.sources.resolve(path, Scope::same())?;
-                    let parsed = parse_file(partial.path())?;
-                    partial.push(parsed);
+                Node::Include(inc_path) => {
+                    let inc_raws = self.resolve_and_ingest(program, inc_path)?;
+                    raws.push(RawOp::Scope(inc_raws));
                 }
-                Node::Include(path) => {
-                    let partial = self.sources.resolve(path, Scope::independent())?;
-                    let parsed = parse_file(partial.path())?;
-                    partial.push(parsed);
-                }
-                Node::IncludeHex(path) => {
-                    let partial = self.sources.resolve(path, Scope::same())?;
-
+                Node::IncludeHex(hex_path) => {
                     let file =
-                        std::fs::read_to_string(partial.path()).with_context(|_| error::Io {
-                            message: "reading hex include",
-                            path: partial.path().to_owned(),
+                        std::fs::read_to_string::<&PathBuf>(&hex_path).with_context(|_| {
+                            error::Io {
+                                message: "reading hex include",
+                                path: <PathBuf as Into<PathBuf>>::into(hex_path.clone()).to_owned(),
+                            }
                         })?;
 
                     let raw = hex::decode(file.trim())
                         .map_err(|e| Box::new(e) as Box<dyn std::error::Error>)
                         .context(error::InvalidHex {
-                            path: partial.path().to_owned(),
+                            path: <PathBuf as Into<PathBuf>>::into(hex_path).to_owned(),
                         })?;
 
-                    partial.push(vec![Node::Raw(raw)]);
+                    raws.push(RawOp::Raw(raw))
                 }
             }
         }
 
-        if !self.sources.sources.is_empty() {
-            panic!("extra sources?");
-        }
+        Ok(raws)
+    }
 
-        Ok(())
+    fn resolve_and_ingest<P: AsRef<Path>>(
+        &mut self,
+        program: &mut Program,
+        path: P,
```

Comment:
Generally in Rust, if you immediately clone a reference inside a function, you may as well take the owned type as a parameter. In this case, `PathBuf`.

That lets the caller decide whether to clone or not. If the caller no longer needs the variable, it can transfer ownership (saving the potentially expensive clone.)

Reply:
I simplified the code a bit so that push_path gets a reference and made some changes to bring the clone to the one place where it was indispensable. Sorry, while I've made a lot of progress in this time, there are still some "common practices" of Rust that escape me.

Reply:
I only mentioned it because of your postscript. It's such a tiny thing that I wouldn't've mentioned it otherwise.

Reply:
Don't worry, these comments are really helpful. There are many things you don't learn in books/tutorials.